addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+                String messageTypeHex = matcher.group(2);
+                String sequence = matcher.group(3);
+                String payload = matcher.group(4);
+                String crc = matcher.group(5);
+
+                if (protocolHeader.equals(PROTOCOL_HEADER)) {
+                    String calculatedCRC = Message.getCRC(messageTypeHex + sequence + payload);
+                    if (calculatedCRC.equals(crc)) {
+                        MessageType messageType = MessageType.forValue(Integer.parseInt(messageTypeHex, 16));
+                        int sequenceNumber = Integer.parseInt(sequence, 16);
+
+                        if (messageType == null) {
+                            logger.debug("Received unrecognized message: messageTypeHex=0x{}, sequence={}, payload={}",
+                                    messageTypeHex, sequenceNumber, payload);
+                            return;
+                        }
+
+                        logger.debug("Received message: messageType={}, sequenceNumber={}, payload={}", messageType,
+                                sequenceNumber, payload);
+
+                        try {
+                            Message message = messageFactory.createMessage(messageType, sequenceNumber, payload);
+
+                            if (message instanceof AcknowledgementMessage
+                                    && !((AcknowledgementMessage) message).isExtended()) {
+                                logger.debug("Adding to acknowledgedQueue: {}", message);
+                                acknowledgedQueue.put((AcknowledgementMessage) message);
+                            } else {
+                                logger.debug("Adding to receivedQueue: {}", message);
+                                receivedQueue.put(message);
+                            }
+                        } catch (IllegalArgumentException e) {
+                            logger.warn("Failed to create message", e);
+                        } catch (InterruptedException e) {
+                            Thread.interrupted();
+                        }
+                    } else {
+                        logger.warn("Plugwise protocol CRC error: {} does not match {} in message", calculatedCRC, crc);
+                    }
+                } else {
+                    logger.debug("Plugwise protocol header error: {} in message {}", protocolHeader, response);
+                }
+            } else if (!response.contains("APSRequestNodeInfo") && !response.contains("APSSetSleepBehaviour")
+                    && !response.startsWith("# ")) {
+                logger.warn("Plugwise protocol message error: {}", response);
+            }
+        }
+    }
+
+    public void removeMessageListener(PlugwiseMessageListener listener) {
+        List<FilteredMessageListener> removedListeners = new ArrayList<>();
+        for (FilteredMessageListener filteredListener : filteredListeners) {
```

Comment:
I don't think that will work because `filteredListeners` is a `CopyOnWriteArrayList` which throws UnsupportedOperationException when using .remove on an iterator:

```
 * the list since the iterator was created.  Element-changing
 * operations on iterators themselves ({@code remove}, {@code set}, and
 * {@code add}) are not supported. These methods throw
 * {@code UnsupportedOperationException}.
```

This code now lives in the PlugwiseFilteredMessageListenerList.